### PR TITLE
Address issue with multiple defines in C++ Comm

### PR DESCRIPTION
**Pull request by chadwalters**
Addresses https://github.com/Microsoft/bond/issues/294
Includes a new example written by Christopher that reproduces the issue.

### CODE REVIEW

**Comment by tstein** [comment]
LGTM. What do you think about putting this in cpp tests instead of examples, though? It's not demonstrating functionality so much as it's proving we don't have a particular bug.

**Comment by chadwalters** [comment]
Our examples tend to provide both -- examples of how to make certain scenarios work and verification that those scenarios compile and function. I agree the scenario seems a bit degenerate (this kind of thing should work out of the box) but I think it still fits better in the context of our examples than our tests.

File: examples/cpp/comm/CMakeLists.txt [hunk 1]
```diff
@@ -4,6 +4,7 @@ add_subdirectory (event)
 add_subdirectory (generic)
 add_subdirectory (layer)
 add_subdirectory (logging)
+add_subdirectory (multi-trans-units)
```
Comment (chwarr) [inline examples/cpp/comm/CMakeLists.txt]: We shouldn't have an explicit multiple translation unit example like this. What I wrote was just to reproduce the reported problem. We should just take one of these existing examples and split it across translation units. E.g., implement a service separate from `main()`.

Comment (chadwalters) [inline examples/cpp/comm/CMakeLists.txt]: Let me just check in the fix itself for now then. We will need to find a good approach for making sure this kind of thing doesn't regress.

**Comment by chadwalters** [comment]
Failure was the usual time out failure on VS12.

**Pull request merged**